### PR TITLE
[misc] Split accessor and other kernels in statistics

### PR DESCRIPTION
**Pull request by xumingkuan**
Related PR = #851

The geometric mean of optimization factor on kernels: 1.1146766978072276

benchmark20200422_2

Tests with significant optimization (> 1.5):
```
test_ad_basics__test_frac : 1.558139534883721
test_ad_basics__test_minmax : 1.5142857142857142
test_ad_basics__test_poly : 1.5714285714285714
test_ad_basics__test_unary : 1.5454545454545454
test_ad_if__test_ad_if : 1.8214285714285714
test_ad_if__test_ad_if_mutable : 2.5671641791044775
test_ad_if__test_ad_if_parallel : 2.357142857142857
test_ad_if__test_ad_if_parallel_complex : 1.82
test_ad_offload__test_offload_order : 1.532258064516129
test_atomic__test_atomic_sub_expr_evaled : 1.6470588235294117
test_basics__test_if_global_load : 1.6428571428571428
test_basics__test_while_global_load : 1.8666666666666667
test_continue__test_kernel_continue : 1.605263157894737
test_continue__test_unconditional_continue : 1.5106382978723405
test_fill__test_fill_matrix_scalar : 1.6
test_numpy_io__test_f64 : 1.5454545454545454
test_numpy_io__test_matrix : 1.5454545454545454
test_numpy_io__test_to_numpy_2d : 1.5454545454545454
test_struct_for__test_nested_2d : 1.5454545454545454
```

Tests that become worse in the sense of the number of statements (< 0.9):
```
test_ndrange__test_static_grouped_static : 0.8478260869565217
test_static__test_static_if_error : 0.8983050847457628
test_types__test_overflow[dt0-8] : 0.8888888888888888
test_types__test_type_tensor64[dt0] : 0.8823529411764706
test_types__test_type_tensor64[dt1] : 0.8888888888888888
test_types__test_type_tensor64[dt2] : 0.8888888888888888
test_types__test_type_tensor[dt1] : 0.8888888888888888
test_types__test_type_tensor[dt2] : 0.8888888888888888
test_types__test_type_tensor[dt3] : 0.8823529411764706
test_types__test_type_tensor[dt4] : 0.8888888888888888
test_types__test_type_tensor[dt5] : 0.8888888888888888
test_types__test_type_tensor[dt6] : 0.8888888888888888
```

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

### CODE REVIEW

**Comment by yuanming-hu** [comment]
Is 
```
auto num_stmts = irpass::analysis::count_statements(this->ir);
if (!ernel->is_accessor)
    stat.add("codegen_kernel_statements", num_stmts);
else
    stat.add("codegen_accessor_statements", num_stmts);
stat.add("codegen_statements", num_stmts);  
```
more informative for your benchmark?

**Comment by yuanming-hu** [comment]
Also, you may want to lower `linearize` even without advanced optimization to make the comparison fairer.

**Comment by xumingkuan** [comment]
> Also, you may want to lower `linearize` even without advanced optimization to make the comparison fairer.

Yes. I'll do that in the next PR.

File: taichi/util/statistics.cpp [hunk 1]
```diff
@@ -28,8 +28,6 @@ void Statistics::print(std::string *output) {
 
 void Statistics::clear() {
   counters.clear();
-  counters["codegen_statements"] = 0;
-  counters["codegen_offloaded_tasks"] = 0;
```
Comment (xumingkuan) [inline taichi/util/statistics.cpp:32]: I'll be using `dict(...).get('codegen_kernel_statements', 0)` so this is no longer necessary.

**Comment by yuanming-hu** [comment]
Out of curiosity, I tested `mpm99.py`, which gives a satisfying 22% improvement: 3549 v.s. 2911.

I also observed some opportunities there:
```
...
      <f32 x1> $242 = mul $120 $241
      <f32*x1> $243 : global store [$119 <- $242] # Not needed because of 253
      <i32 x1> $244 = global load $127
      <i32 x1> $245 = cmp_eq $244 $196
      <i32 x1> $246 = bit_and $34 $245
      <f32 x1> $247 = max $214 $236
      <f32 x1> $248 = min $247 $238
      <f32 x1> $249 = select($246, $248, $214)
      <f32 x1> $250 = global load $119 # Not needed since it's 242
      <f32 x1> $251 = div $214 $249
      <f32 x1> $252 = mul $250 $251
      <f32*x1> $253 : global store [$119 <- $252]
...


```

**Comment by xumingkuan** [comment]
> Out of curiosity, I tested `mpm99.py`, which gives a satisfying 22% improvement: 3549 v.s. 2911.
> 
> I also observed some opportunities there:
> 
> ```
> ...
>       <f32 x1> $242 = mul $120 $241
>       <f32*x1> $243 : global store [$119 <- $242] # Not needed because of 253
>       <i32 x1> $244 = global load $127
>       <i32 x1> $245 = cmp_eq $244 $196
>       <i32 x1> $246 = bit_and $34 $245
>       <f32 x1> $247 = max $214 $236
>       <f32 x1> $248 = min $247 $238
>       <f32 x1> $249 = select($246, $248, $214)
>       <f32 x1> $250 = global load $119 # Not needed since it's 242
>       <f32 x1> $251 = div $214 $249
>       <f32 x1> $252 = mul $250 $251
>       <f32*x1> $253 : global store [$119 <- $252]
> ...
> ```

Great! Maybe I'll copy `optimize_local_variable.cpp` to `optimize_global_variable.cpp` tomorrow, for both global temps and global pointers. The logic would be the same (but current infrastructure would require copying the code and change all `local` to `global`s rather than reuse some functions).

File: taichi/util/statistics.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/util/statistics.cpp:32]: OK sounds good

**Pull request merged**